Implementation for issue 601 Google Cloud Storage plugin

wshands opened the pull request:
Plugin implementation for access to files stored in Google Cloud Storage

Commit:
initial checkin of GCS Plugin code

Commit:
add metadata on upload and fix up code

Commit:
clean up code

Commit:
upload and download in chunks

Commit:
fix null content type on Mac

Commit:
make sure destination paths are created for a WDL workflow or tool

Commit:
close writer so file appears in GCS

Commit:
close writer so file appears in GCS

Commit:
remove unnecessary initializer

Commit:
remove unnecessary variable

Commit:
remove unnecessary variable

Commit:
remove unnecessary variable

coverbeck commented on pom.xml:
```diff
@@ -0,0 +1,231 @@
+        <github.url>scm:git:git@github.com:dockstore/gs-plugin.git</github.url>
+    </properties>
+
+    <distributionManagement>
```
I don't think you need this section, since we will be doing GitHub releases, which is different than how s3 plugin is done.

wshands replied:
The distribution management section?

coverbeck replied:
Yes, the whole distributionManagement section should not be necessary.

wshands replied:
Removed distributionManagment section

coverbeck commented on pom.xml:
```diff
@@ -0,0 +1,231 @@
+        <dependency>
+            <groupId>io.dockstore</groupId>
+            <artifactId>dockstore-file-plugin-parent</artifactId>
+            <version>1.4.5</version>
```
Please set to the latest public version, which should 1.5.1

wshands replied:
done

coverbeck commented on src/main/java/io/dockstore/provision/GSPlugin.java:
```diff
@@ -0,0 +1,271 @@
+        public boolean downloadFrom(String sourcePath, Path destination) {
+            Storage gsClient = getGoogleGSClient();
+
+            String trimmedPath = sourcePath.replace("gs://", "");
```
I would pull this out into a function and write a unit test for it.

wshands replied:
Refactored to separate method and test added

coverbeck commented on src/main/java/io/dockstore/provision/GSPlugin.java:
```diff
@@ -0,0 +1,271 @@
+
+            Storage gsClient = getGoogleGSClient();
+
+            String trimmedPath = destPath.replace("gs://", "");
```
Would put this in a unit tested function.

wshands replied:
Refactored to separate method and test added

coverbeck commented on src/main/java/io/dockstore/provision/GSPlugin.java:
```diff
@@ -0,0 +1,271 @@
+                    return false;
+                }
+                // create the blob in one request.
+                gsClient.create(blobInfo, bytes);
```
What happens when this fails? Does it throw an exception or do you need to check the return value? Seems like you need to handle that case.

wshands replied:
try catch added

Commit:
remove unnecessary variable

coverbeck commented on src/main/java/io/dockstore/provision/GSPlugin.java:
```diff
@@ -0,0 +1,271 @@
+                try {
+                    Map<String, String> map = gson.fromJson(metadata.get(), type);
+                    for (Map.Entry<String, String> entry : map.entrySet()) {
+                        System.out.println("Loading " + entry.getKey() + "->" + entry.getValue());
```
I don't think you should be printing this. Or is there a reason you are doing so beyond debugging?

wshands replied:
Commented out; should there be a logger?

coverbeck replied:
We don't think logging works, but there is a ticket to investigate, ga4gh/dockstore#1483

wshands replied:
I have left logging implementation for the other ticket

coverbeck commented on src/main/java/io/dockstore/provision/GSPlugin.java:
```diff
@@ -0,0 +1,271 @@
+                } catch (com.google.gson.JsonSyntaxException ex) {
+                    System.err.println(
+                            "Could not load metadata. Metadata syntax exception. Uploading file without metadata:" + ex.getMessage());
+                    blobInfo = BlobInfo.newBuilder(blobId).setContentType(contentType).build();
```
One could argue this both ways, but I would lean towards having the upload fail if the metadata is not correct. If you are trying to set the metadata and it fails, then I would think you would not want to proceed.

wshands replied:
Hmm...just left it as is

wshands replied:
Changed code to fail upload if metadata cannot be set

coverbeck commented on src/main/java/io/dockstore/provision/GSPlugin.java:
```diff
@@ -0,0 +1,271 @@
+
+            // Bug on Mac where Files.probeContentType returns null
+            if (contentType == null)
+                contentType = "application/octet-stream";
```
String is duplicated.

wshands replied:
I have to set it again because it gets set to null on a Mac when Files.probeContentType is called

coverbeck replied:
You have the string "application/octet-stream" twice. What if you make a typo in one of the instances? You should define a variable that has the string once, and use the variable twice.

wshands replied:
Added a static string variable to hold the content type

Commit:
add try catch around blob create
comment out metadata print statement
remove pom distributionManagement section
bump dockstore file plugin parent version

Commit:
use define to hold the default content type string

denis-yuen commented on README.md:
```diff
@@ -1,2 +1,36 @@
+{
+  "input_file": {
+        "class": "File",
+        "path": "gs://oicr.temp/bamstats_report.zip"
```
Should probably note that these aren't real paths and won't work (unless they now are? in which case, maybe they shouldn't be in a "temp" directory)

wshands replied:
Added comment that the URLs do not necessarily exist. 
Do we have a public OICR GCS URL where we can put example input files?

denis-yuen replied:
Nope, could be an idea to add one. 
Or maybe there are already public examples of files we can use?

wshands replied:
I'll take a look to see if I can find a good public URL

denis-yuen commented on pom.xml:
```diff
@@ -0,0 +1,239 @@
+
+    <licenses>
+        <license>
+            <name>GNU General Public License (GPLv3)</name>
```
This was probably a bad cut-and-paste on my part, we should be using Apache 2.0 for new code
(or updates of old code)

wshands replied:
done

denis-yuen commented on src/main/java/io/dockstore/provision/GSPlugin.java:
```diff
@@ -0,0 +1,300 @@
+        public boolean downloadFrom(String sourcePath, Path destination) {
+            Storage gsClient = getGoogleGSClient();
+
+            //String trimmedPath = sourcePath.replace("gs://", "");
```
Unless there's a good chance commented code will be uncommented (and there should be a comment why), don't commit commented code (good summary of my general train of thought https://blog.kentcdodds.com/please-don-t-commit-commented-out-code-53d0b5b26d5f )

wshands replied:
removed commented code

denis-yuen commented on src/main/java/io/dockstore/provision/GSPlugin.java:
```diff
@@ -0,0 +1,300 @@
+                }
+            }
+
+            writeTo.close();
```
Would be cleaner with a try catch with resources

wshands replied:
added try catch with resources

denis-yuen commented on src/test/java/io/dockstore/provision/GSPluginUnitTest.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+
+//    @Test
```
Same here (these look useful?)

wshands replied:
removed commented tests; will add these as integration tests later?

denis-yuen reviewed (changes requested):
Some minor cleanup

Commit:
use define to hold the default content type string

Commit:
Add file names to error messages
Add more try catch blocks for specific exceptions

Commit:
Add file names to error messages
Add more try catch blocks for specific exceptions

Commit:
Add file names to error messages
Add more try catch blocks for specific exceptions

Commit:
fix

Commit:
fix

Commit:
improve exception handling and error messages

Commit:
provide a valid input GCS path

coverbeck commented on README.md:
````diff
@@ -1,2 +1,51 @@
+    }
+}
+```
+Note that metadata is Base64 encoded in the JSON and creates metadata tags on the uploaded file.
````
I realize this line is copied from the s3 plugin, but I just noticed it. I find "in the JSON" confusing; it sounds like it could be talking about the `test.gs.json` file, but it also needs to be clear that the value itself is a Base64-encoded JSON. 

It might also be nice to have an example JSON, and how you BASE64 encode it, but that may be overkill until people start using this feature.

denis-yuen replied:
There are a couple examples here that can be copied/modified
https://github.com/ga4gh/dockstore/search?l=JSON&q=metadata

wshands replied:
Provided an example of encoding the metadata and reworded the Note

coverbeck reviewed (approved):
I made a suggestion, but given that it's already in the s3 plugin like that, it's optional.

Commit:
provide a valid input GCS path

Commit:
Fail the upload if metadata is not structured properly
Reformat to 100 chars per line

denis-yuen commented on README.md:
````diff
@@ -1,2 +1,56 @@
+    }
+}
+```
+Note that metadata is Base64 encoded JSON and creates metadata tags on the uploaded file.
````
Might want to link to https://cloud.google.com/storage/docs/gsutil/addlhelp/WorkingWithObjectMetadata#custom-metadata and provide the gsutil equivalent of what the metadata in the test.gs.json does

wshands replied:
done

denis-yuen commented on pom.xml:
```diff
@@ -0,0 +1,239 @@
+                <artifactId>maven-release-plugin</artifactId>
+                <version>2.5.3</version>
+            </plugin>
+            <!-- DOESN'T WORK WITH MAVEN 3 (I defined the plugin metadata in properties section)
```
Can delete if not relevant

wshands replied:
deleted

denis-yuen commented on src/main/java/io/dockstore/provision/GSPlugin.java:
```diff
@@ -0,0 +1,330 @@
+        private static final int MAX_FILE_SIZE_FOR_SINGLE_WRITE = 1_000_000;
+        private static final int DOWNLOAD_BUFFER_SIZE = 64 * 1024;
+        private static final int UPLOAD_BUFFER_SIZE = 1024;
+        private static final String DEFAULT_CONTENT_TYPE = "application/octet-stream";
```
Can probably just use https://docs.oracle.com/javaee/7/api/javax/ws/rs/core/MediaType.html#APPLICATION_OCTET_STREAM

wshands replied:
done (used com.google.common.net.MediaType)

denis-yuen commented on src/main/java/io/dockstore/provision/GSPlugin.java:
```diff
@@ -0,0 +1,330 @@
+            String bucketName = getBucketName(sourcePath);
+            String blobName = getBlobName(sourcePath);
+
+            //Blob blobMetadata = gsClient.get(bucketName, blobName, Storage.BlobGetOption.fields(Storage.BlobField.values()));
```
Delete if not likely to be uncommented

wshands replied:
done

denis-yuen commented on src/main/java/io/dockstore/provision/GSPlugin.java:
```diff
@@ -0,0 +1,330 @@
+                return false;
+            }
+
+            ProgressPrinter printer = new ProgressPrinter();
```
Don't think these (printer, fileContent, writeTo) get closed on exception
Note that a single try ... catch with resources can handle multiple resources so that might make things easier/cleaner

wshands replied:
done for fileContent and writeTo but ProgressPrinter doesn't have a close method and does not support auto-close?

denis-yuen replied:
Ah, ok

denis-yuen commented on src/main/java/io/dockstore/provision/GSPlugin.java:
```diff
@@ -0,0 +1,330 @@
+
+            } else {
+                // When Blob size is big or unknown use the blob's channel reader.
+                try (ReadChannel reader = blob.reader()) {
```
Read will be closed properly here for example

denis-yuen reviewed (approved):
Think we're in the right neighborhood, a few minor revisions

Commit:
Added URL to information about metadata
Used try catch with resources for fileContent and writeTo
Moved ProgressPrinter close to code where used

Commit:
used com.google.common.net.MediaType for contentType default
removed commented code from pom.xml
